src/Renci.SshNet/ShellStream.cs

```diff
@@ -248,35 +263,40 @@ public void Expect(params ExpectAction[] expectActions)
         public void Expect(TimeSpan timeout, params ExpectAction[] expectActions)
         {
             var expectedFound = false;
-            var text = string.Empty;
+            var matchText = string.Empty;
 
             do
             {
                 lock (_incoming)
                 {
-                    if (_incoming.Count > 0)
+                    if (_expect.Count > 0)
                     {
-                        text = _encoding.GetString(_incoming.ToArray(), 0, _incoming.Count);
+                        matchText = _encoding.GetString(_expect.ToArray(), 0, _expect.Count);
                     }
 
-                    if (text.Length > 0)
+                    if (matchText.Length > 0)
                     {
                         foreach (var expectAction in expectActions)
                         {
-                            var match = expectAction.Expect.Match(text);
+                            var match = expectAction.Expect.Match(matchText);
 
                             if (match.Success)
                             {
-                                var result = text.Substring(0, match.Index + match.Length);
-                                var charCount = _encoding.GetByteCount(result);
+                                var returnText = matchText.Substring(0, match.Index + match.Length);
+                                var returnLength = _encoding.GetByteCount(returnText);
 
-                                for (var i = 0; i < charCount && _incoming.Count > 0; i++)
+                                // Remove processed items from the queue
+                                for (var i = 0; i < returnLength && _incoming.Count > 0; i++)
```

Comment:
Is this actually going to remove what it should from both `_incoming` and `_expect`?

If `_incoming` looks like: `aaaaabbbbbcccccZ`
and `_expect` (and so `matchText`) looks like: `cccZ`
and we are expecting `Z`
then `returnText == "cccZ"` and `returnLength == 4`

Then we are going to dequeue `aaaa` from `_incoming` and nothing from `_expect` (because _incoming.Count > _expect.Count + 4). So we end up with:

`_incoming` looks like: `abbbbbcccccZ`
and `_expect` still looks like: `cccZ`

Have I got that right? Is that expected? (it feels wrong)

Reply:
I'm going to add a test to try to replicate this and make sure its accounted for. I have a feeling that the dequeueing may be slightly off since for things to happen as you've mentioned, data would have to accumulate and be read in different ways within the same workflow.